aptos-move/aptos-vm/src/move_vm_ext/vm.rs

```diff
@@ -1,175 +1,186 @@
+            MiscGasParameters::zeros(),
+            timed_features.clone(),
+            features.clone(),
+        );
+
+        let vm = MoveVM::new_with_config(
+            aptos_natives_with_builder(&mut native_builder),
+            vm_config.clone(),
+        );
+
+        Self {
+            vm,
+            chain_id,
+            features,
+        }
+    }
+
+    pub fn genesis_change_set_configs(&self) -> ChangeSetConfigs {
+        // Because genesis sessions are not metered, there are no change set
+        // (storage) costs as well.
+        ChangeSetConfigs::unlimited_at_gas_feature_version(LATEST_GAS_FEATURE_VERSION)
+    }
+
+    pub fn new_genesis_session<'r, R: AptosMoveResolver>(
+        &self,
+        resolver: &'r R,
+        genesis_id: HashValue,
+    ) -> SessionExt<'r, '_> {
+        let session_id = SessionId::genesis(genesis_id);
+        SessionExt::new(
+            session_id,
+            &self.vm,
+            self.chain_id,
+            &self.features,
+            None,
+            resolver,
+        )
+    }
+}
 
 pub struct MoveVmExt {
     inner: MoveVM,
-    chain_id: u8,
-    features: Features,
+    pub(crate) env: Arc<Environment>,
 }
 
 impl MoveVmExt {
     fn new_impl<F>(
         native_gas_params: NativeGasParameters,
         misc_gas_params: MiscGasParameters,
         gas_feature_version: u64,
-        chain_id: u8,
-        features: Features,
-        timed_features: TimedFeatures,
+        env: Arc<Environment>,
         gas_hook: Option<F>,
         resolver: &impl AptosMoveResolver,
-        aggregator_v2_type_tagging: bool,
-    ) -> VMResult<Self>
+    ) -> Self
     where
         F: Fn(DynamicExpression) + Send + Sync + 'static,
     {
         let mut builder = SafeNativeBuilder::new(
             gas_feature_version,
             native_gas_params.clone(),
             misc_gas_params.clone(),
-            timed_features.clone(),
-            features.clone(),
+            env.timed_features.clone(),
+            env.features.clone(),
         );
         if let Some(hook) = gas_hook {
             builder.set_gas_hook(hook);
         }
 
-        let paranoid_type_checks = crate::AptosVM::get_paranoid_checks();
-        let vm_config = aptos_prod_vm_config(
-            &features,
-            &timed_features,
-            aggregator_v2_type_tagging,
-            paranoid_type_checks,
-        );
-
-        Ok(Self {
-            inner: WarmVmCache::get_warm_vm(builder, vm_config, resolver)?,
-            chain_id,
-            features,
-        })
+        Self {
+            inner: WarmVmCache::get_warm_vm(
+                builder,
+                env.vm_config.clone(),
+                resolver,
+                env.features.is_enabled(FeatureFlag::VM_BINARY_FORMAT_V7),
+            )
+            .expect("should be able to create Move VM; check if there are duplicated natives"),
+            env,
+        }
     }
 
     pub fn new(
```

Comment:
how often is new used? I feel like it might be nicer to combine new, new_impl and new_with_gas_hook all in a single new API with optional gas hook that does what impl does right now.

Reply:
New is used a lot, in tests mostly (unlike new_with_gas_hook which is only called once), combining APIs will result in passing Nones everywhere. I would say we need to fix tests first?

In particular, having gas params not part of an environment is pretty bad, I think there is a nice ay, and will do it in the next PR.